Gh-172: Document the Job Tracker (#458)
gchq/gaffer-doc: docs/administration-guide/job-tracker.md

# Job Tracker

This section covers how to configure your Gaffer graph to allow jobs to be executed.

'Job' refers to an operation chain with a long-running sequence of operations that is executed
asynchronously. 
Jobs are, therefore, useful for operation chains which take a long time to return results or if 
you want to cache the results of your operation chain. 

Jobs are only available when the job tracker has been enabled. 
The job tracker can then be used to store the status of current and historic jobs 
while a 'cache' is used to store any job results. 

## Configuration

By default, the job tracker is disabled. 

> It can be enabled by setting its associated [store property](../administration-guide/gaffer-stores/store-guide.md/#cache-service) to true.

Reviewer: This is linking to the cache section, I think it should link to the properties list which is where the option to enable job tracker is.
```suggestion
It can be enabled by setting its associated [store property](../administration-guide/gaffer-stores/store-guide.md#all-general-store-properties) to true.
```

There must also be a [cache configured](../administration-guide/gaffer-stores/store-guide.md/#caches) for the job
tracker to work.

> To store any results using the job tracker, a `GafferResultCache` needs to enabled. 

Reviewer: We could have a subheading before this section about the results cache starts.

This job results cache can then store your results using operations and operation handlers 
(these are disabled by default). 

> Note that this is note the same as your initial cache, as this results 'cache' is simply a second Gaffer graph

Reviewer: ```suggestion
Note that this is not the same as your initial cache, as this results 'cache' is simply a second Gaffer graph
```

Together these allow you to store your results in a seperate Gaffer graph, your 'cache'.

!!! example "Adding the operations to your Operations Declarations file"
    
    ``` json
    {
        "operations": [
            {
                "operation": "uk.gov.gchq.gaffer.operation.impl.export.resultcache.ExportToGafferResultCache",
                "handler": {
                    "class": "uk.gov.gchq.gaffer.operation.export.resultcache.handler.ExportToGafferResultCacheHandler",
                    "graphId": "resultCacheGraph",
                    "timeToLive": 86400000,
                    "storePropertiesPath": "cache-store.properties"
                }
            },
            {
                "operation": "uk.gov.gchq.gaffer.operation.impl.export.resultcache.GetGafferResultCacheExport",
                "handler": {
                    "class": "uk.gov.gchq.gaffer.operation.export.resultcache.handler.GetGafferResultCacheExportHandler",
                    "graphId": "resultCacheGraph",
                    "timeToLive": 86400000,
                    "storePropertiesPath": "cache-store.properties"
                }
            }
        ]
    }
    ```

> As the results 'cache' is a separate Gaffer graph; an additional `cache-store.properties` file will need to be created
> which contains the properties for your results graph. 

Reviewer: ```suggestion
As the results 'cache' is a separate Gaffer graph; an additional `cache-store.properties` file must be created to
configure the store properties for your results graph. 
```

asynchronously. 
Jobs are, therefore, useful for operation chains which take a long time to return results or if 
you want to cache the results of your operation chain. 

Jobs are only available when the job tracker has been enabled. 
The job tracker can then be used to store the status of current and historic jobs 
while a 'cache' is used to store any job results. 

## Configuration

By default, the job tracker is disabled. 
It can be enabled by setting its associated [store property](../administration-guide/gaffer-stores/store-guide#store-properties) to true.
There must also be a [cache configured](../administration-guide/gaffer-stores/store-guide.md#caches) for the job
tracker to work.

### Configuring the Results Cache
To store any results using the job tracker, a `GafferResultCache` needs to enabled. 
This job results cache can then store your results using operations and operation handlers 
(these are disabled by default). 
Note that this is not the same as your initial cache, as this results 'cache' is simply a second Gaffer graph
which will store your results.

To store results using the job tracker a few elements must be set up. 

> Firstly, you need to add two operations to an [operations declarations JSON file](../development-guide/example-deployment/project-setup.md#operations-declarations). 

Reviewer 2: Might be better to link to the configuration page instead https://gchq.github.io/gaffer-doc/latest/administration-guide/gaffer-config/config/#operationsdeclarationsjson

!!! example "Adding the operations to your Operations Declarations file"
    
    ``` json
    {
        "operations": [
            {
                "operation": "uk.gov.gchq.gaffer.operation.impl.export.resultcache.ExportToGafferResultCache",
                "handler": {
                    "class": "uk.gov.gchq.gaffer.operation.export.resultcache.handler.ExportToGafferResultCacheHandler",
                    "graphId": "resultCacheGraph",
                    "timeToLive": 86400000,
                    "storePropertiesPath": "cache-store.properties"
                }
            },
            {
                "operation": "uk.gov.gchq.gaffer.operation.impl.export.resultcache.GetGafferResultCacheExport",
                "handler": {
                    "class": "uk.gov.gchq.gaffer.operation.export.resultcache.handler.GetGafferResultCacheExportHandler",
                    "graphId": "resultCacheGraph",
                    "timeToLive": 86400000,
                    "storePropertiesPath": "cache-store.properties"
                }
            }
        ]
    }
    ```

As the results 'cache' is a separate Gaffer graph; an additional `cache-store.properties` file must be created

> to configure store the properties for your results graph. 

Reviewer: ```suggestion
to configure the store properties for your results graph. 
```
